File: turborepo-tests/tui/tests/tui.test.mjs
import assert from "node:assert/strict"
import { cp, mkdir, mkdtemp, rm, writeFile } from "node:fs/promises"
import { tmpdir } from "node:os"
import { dirname, join, resolve } from "node:path"
import { fileURLToPath } from "node:url"
import test from "node:test"

import { TerminalControl } from "@kitlangton/terminal-control"

const packageDirectory = resolve(dirname(fileURLToPath(import.meta.url)), "..")
const repositoryRoot = resolve(packageDirectory, "../..")
const fixtureSource = join(packageDirectory, "fixtures/basic")
const turboBinary = process.env.TURBO_BINARY_PATH ?? join(repositoryRoot, "target/debug/turbo")
const startupTimeout = 15_000
// Known failing regressions stay skipped in CI but remain directly reproducible.
const runKnownBugs = process.env.TURBO_TEST_KNOWN_BUGS === "1"

function inheritedEnvironment() {
  const environment = {
    PATH: process.env.PATH,
    HOME: process.env.HOME,
    LANG: process.env.LANG,
    SHELL: process.env.SHELL,
    TMPDIR: process.env.TMPDIR,
    USER: process.env.USER,
    COREPACK_ENABLE_DOWNLOAD_PROMPT: "0",
    TURBO_GLOBAL_WARNING_DISABLED: "1",
    TURBO_PRINT_VERSION_DISABLED: "1",
    TURBO_TELEMETRY_MESSAGE_DISABLED: "1",
    DO_NOT_TRACK: "1",
    NPM_CONFIG_UPDATE_NOTIFIER: "false",
  }

  return Object.fromEntries(Object.entries(environment).filter(([, value]) => value !== undefined))
}

async function launchTui({
  extraArgs = [],
  prepareWorkspace,
  task = "dev",
  viewport = { cols: 100, rows: 30 },
} = {}) {
  const directory = await mkdtemp(join(tmpdir(), "turbo-tui-test-"))
  const workspace = join(directory, "workspace")
  await cp(fixtureSource, workspace, { recursive: true })
  await prepareWorkspace?.(workspace)

  const configDirectory = join(directory, "config")
  await mkdir(configDirectory)

  let terminal
  try {
    terminal = await TerminalControl.make()
    const session = await terminal.launch({
      command: [turboBinary, "run", task, "--ui=tui", "--skip-infer", ...extraArgs],
      cwd: workspace,
      viewport,
      inheritEnv: false,
      env: {
        ...inheritedEnvironment(),
        TURBO_CONFIG_DIR_PATH: configDirectory,
      },
    })

    return { directory, session, terminal }
  } catch (error) {
    await terminal?.close().catch(() => {})
    await rm(directory, { force: true, recursive: true })
    throw error
  }
}

async function closeTui(context) {
  try {
    await context.session.stop()
  } finally {
    try {
      await context.terminal.close()
    } finally {
      await rm(context.directory, { force: true, recursive: true })
    }
  }
}

async function waitForScreen(session, description, predicate, timeoutMs = startupTimeout) {
  try {
    return await session.screen.waitUntil(
      (screen) => predicate(screen.text, screen.frame),
      { timeoutMs },
    )
  } catch (error) {
    const screen = await session.screen.capture({ allowIncomplete: true, deadlineMs: 0, settleMs: 0 })
    throw new Error(`${description}\n\nVisible screen:\n${screen.text}`, { cause: error })
  }
}

async function addOverflowingTaskList(workspace) {
  await writeFile(
    join(workspace, "turbo.json"),
    JSON.stringify({
      $schema: "https://turborepo.dev/schema.json",
      ui: "tui",
      tasks: { list: { cache: false } },
    }),
  )

  for (let index = 0; index < 20; index++) {
    const name = `zz-task-${String(index).padStart(2, "0")}`
    const packageDirectory = join(workspace, "packages", name)
    await mkdir(packageDirectory)
    await writeFile(
      join(packageDirectory, "package.json"),
      JSON.stringify({
        name,
        private: true,
        scripts: { list: "node -e \"setTimeout(() => {}, 30000)\"" },
      }),
    )
  }
}

async function waitForTranscriptIdle(session, quietForMs = 500, timeoutMs = startupTimeout) {
  const deadline = Date.now() + timeoutMs
  let transcript = await session.transcript.ansi()
  let quietSince = Date.now()

  while (Date.now() < deadline) {
    await new Promise((resolve) => setTimeout(resolve, 50))
    const next = await session.transcript.ansi()
    if (next.byteLength !== transcript.byteLength) {
      transcript = next
      quietSince = Date.now()
    } else if (Date.now() - quietSince >= quietForMs) {
      return next
    }
  }

  throw new Error("terminal output did not become idle")
}

function visibleSidebarTasks(text) {
  return text.split("\n").flatMap((line) => {
    const divider = line.indexOf("│")
    if (divider < 0) return []
    return line.slice(0, divider).match(/[a-z0-9-]+#list/g) ?? []
  })
}

async function waitForInitialScreen(session) {
  return waitForScreen(session, "TUI did not render its initial task view", (text) =>
    text.includes("Tasks (/ - Search)") &&
    text.includes("interactive#dev") &&
    text.includes("finite#dev") &&
    text.includes("INTERACTIVE_READY")
  )
}

test("renders tasks and navigates between their output", { timeout: 30_000 }, async () => {
  const context = await launchTui()
  try {
    await waitForInitialScreen(context.session)

    await context.session.keyboard.press("ArrowDown")
    await waitForScreen(context.session, "finite task output was not selected", (text) =>
      text.includes("FINITE_COMPLETE")
    )

    await context.session.keyboard.press("ArrowUp")
    await waitForScreen(context.session, "interactive task output was not reselected", (text) =>
      text.includes("INTERACTIVE_READY")
    )
  } finally {
    await closeTui(context)
  }
})

test("drives search and popup overlays", { timeout: 30_000 }, async () => {
  const context = await launchTui()
  try {
    await waitForInitialScreen(context.session)

    await context.session.keyboard.type("/finite")
    await context.session.keyboard.press("Enter")
    await waitForScreen(context.session, "search did not select the finite task", (text) =>
      text.includes("/ finite") && text.includes("FINITE_COMPLETE")
    )

    await context.session.keyboard.press("Escape")
    await context.session.keyboard.type("m")
    await waitForScreen(context.session, "keybind popup did not open", (text) =>
      text.includes("Keybinds") && text.includes("Toggle log panel")
    )

    await context.session.keyboard.press("Escape")
    await context.session.keyboard.type("l")
    await waitForScreen(context.session, "log panel did not open", (text) => text.includes("l to close"))
    await context.session.keyboard.press("Escape")
  } finally {
    await closeTui(context)
  }
})

test("forwards input to an interactive task and returns to navigation", { timeout: 30_000 }, async () => {
  const context = await launchTui()
  try {
    await waitForInitialScreen(context.session)

    await context.session.keyboard.type("i")
    await waitForScreen(context.session, "interactive mode did not start", (text) =>
      text.includes("Ctrl-z - Stop interacting")
    )

    await context.session.keyboard.type("hello-from-test")
    await context.session.keyboard.press("Enter")
    await waitForScreen(context.session, "task did not receive interactive input", (text) =>
      text.includes("INTERACTIVE_ECHO:hello-from-test")
    )

    await context.session.keyboard.press("Control+Z")
    await waitForScreen(context.session, "interactive mode did not return to navigation", (text) =>
      text.includes("Tasks (/ - Search)") && !text.includes("Ctrl-z - Stop interacting")
    )
  } finally {
    await closeTui(context)
  }
})

test("streams logs and restores the terminal on shutdown", { timeout: 30_000 }, async () => {
  const context = await launchTui()
  try {
    await waitForInitialScreen(context.session)

    await context.session.keyboard.type("h")
    await waitForScreen(context.session, "selected task logs did not start streaming", (text) =>
      text.includes("Streaming logs for interactive#dev. Press h to return to the TUI.") &&
      text.includes("INTERACTIVE_READY")
    )

    await context.session.keyboard.type("h")
    await waitForScreen(context.session, "TUI did not return after streaming", (text) =>
      text.includes("INTERACTIVE_READY") && !text.includes("Streaming logs for interactive#dev")
    )

    await context.session.keyboard.press("Control+C")
    const exit = await context.session.waitForExit({ timeoutMs: 10_000 })
    assert.equal(exit.reason, "exited")

    const restored = await context.session.screen.capture({
      allowIncomplete: true,
      deadlineMs: 0,
      settleMs: 0,
    })
    assert.doesNotMatch(restored.text, /Tasks \(\/ - Search\)/)
  } finally {
    await closeTui(context)
  }
})

test(
  "scrolls the task list with the mouse wheel over the sidebar",
  { skip: runKnownBugs ? false : "known bug", timeout: 30_000 },
  async () => {
    const context = await launchTui({
      extraArgs: ["--concurrency=10"],
      prepareWorkspace: addOverflowingTaskList,
      task: "list",
      viewport: { cols: 100, rows: 15 },
    })
    try {
      await waitForScreen(context.session, "overflowing task list did not render", (text) =>
        text.includes("Tasks (/ - Search)") && visibleSidebarTasks(text).length >= 8
      )

      await waitForTranscriptIdle(context.session)
      const beforeScroll = await context.session.screen.capture({
        allowIncomplete: true,
        deadlineMs: 0,
        settleMs: 0,
      })
      const beforeTasks = visibleSidebarTasks(beforeScroll.text)

      const scrollDownOverSidebar = new TextEncoder().encode("\x1b[<65;5;5M".repeat(5))
      await context.session.keyboard.write(scrollDownOverSidebar)
      await new Promise((resolve) => setTimeout(resolve, 500))

      const afterScroll = await context.session.screen.capture({
        allowIncomplete: true,
        deadlineMs: 0,
        settleMs: 0,
      })
      assert.notDeepEqual(
        visibleSidebarTasks(afterScroll.text),
        beforeTasks,
        "mouse-wheel input over the sidebar did not scroll the task list",
      )
    } finally {
      await closeTui(context)
    }
  },
)

test(
  "repaints the TUI immediately after a terminal resize",
  { timeout: 30_000 },
  async () => {
    const context = await launchTui()
    try {
      await waitForInitialScreen(context.session)
      const beforeResize = await waitForTranscriptIdle(context.session)

      for (let step = 0; step <= 10; step++) {
        await context.session.resize({ cols: 110 + step, rows: 26 + step })
        await new Promise((resolve) => setTimeout(resolve, 10))
      }
      await new Promise((resolve) => setTimeout(resolve, 750))

      const afterResize = await context.session.transcript.ansi()
      const repaint = afterResize.slice(beforeResize.byteLength)
      assert.ok(repaint.byteLength > 0, "Turbo emitted no terminal output in response to the resize")
      assert.ok(new TextDecoder().decode(repaint).includes("\x1b[2J"), "Turbo did not clear and repaint the resized terminal")

      const screen = await context.session.screen.capture({
        allowIncomplete: true,
        deadlineMs: 0,
        settleMs: 0,
      })
      assert.equal(screen.frame.cols, 120)
      assert.equal(screen.frame.rows, 36)
    } finally {
      await closeTui(context)
    }
  },
)

File: crates/turborepo/tests/tui_test.rs
#![cfg(target_os = "linux")]
#![cfg_attr(test, allow(clippy::expect_used, clippy::unwrap_used))]

mod common;

use std::{
    collections::BTreeMap,
    fs,
    path::Path,
    thread,
    time::{Duration, Instant},
};

use anyhow::{Context, Result, bail};
use serde_json::json;
use tempfile::TempDir;
use terminal_control::{
    frame::{Cell, Color, Frame},
    session::Session,
    shot::{Options, Shot},
};

const STARTUP_TIMEOUT: Duration = Duration::from_secs(15);

struct TuiContext {
    session: Session,
    _directory: TempDir,
}

fn launch_tui() -> Result<TuiContext> {
    launch_tui_with("dev", &[], 100, 30, |_| Ok(()))
}

fn launch_tui_with(
    task: &str,
    extra_args: &[&str],
    cols: u16,
    rows: u16,
    prepare: impl FnOnce(&Path) -> Result<()>,
) -> Result<TuiContext> {
    let directory = tempfile::tempdir()?;
    common::setup::copy_fixture("tui", directory.path())?;
    common::setup::setup_git(directory.path())?;
    prepare(directory.path())?;

    let config_dir = directory.path().join("config");
    fs::create_dir(&config_dir)?;

    let mut command = vec![
        assert_cmd::cargo::cargo_bin("turbo")
            .to_string_lossy()
            .into_owned(),
        "run".to_owned(),
        task.to_owned(),
        "--ui=tui".to_owned(),
        "--skip-infer".to_owned(),
    ];
    command.extend(extra_args.iter().map(|arg| (*arg).to_owned()));

    let mut env = BTreeMap::from([
        ("COREPACK_ENABLE_DOWNLOAD_PROMPT".to_owned(), "0".to_owned()),
        ("DO_NOT_TRACK".to_owned(), "1".to_owned()),
        ("NPM_CONFIG_UPDATE_NOTIFIER".to_owned(), "false".to_owned()),
        (
            "TURBO_CONFIG_DIR_PATH".to_owned(),
            config_dir.to_string_lossy().into_owned(),
        ),
        ("TURBO_GLOBAL_WARNING_DISABLED".to_owned(), "1".to_owned()),
        ("TURBO_PRINT_VERSION_DISABLED".to_owned(), "1".to_owned()),
        (
            "TURBO_TELEMETRY_MESSAGE_DISABLED".to_owned(),
            "1".to_owned(),
        ),
    ]);
    for key in ["HOME", "LANG", "PATH", "SHELL", "TMPDIR", "USER"] {
        if let Ok(value) = std::env::var(key) {
            env.insert(key.to_owned(), value);
        }
    }

    let session = Session::start(
        &command,
        Some(directory.path()),
        None,
        &Options {
            cols,
            rows,
            env,
            inherit_env: false,
            ..Options::default()
        },
    )?;

    Ok(TuiContext {
        session,
        _directory: directory,
    })
}

fn capture(session: &mut Session) -> Result<Shot> {
    Ok(session.capture(Duration::ZERO, Duration::ZERO)?.shot)
}

fn wait_for_screen(
    session: &mut Session,
    description: &str,
    timeout: Duration,
    mut predicate: impl FnMut(&str, &Frame) -> bool,
) -> Result<Shot> {
    let deadline = Instant::now() + timeout;
    loop {
        let shot = capture(session)?;
        let text = shot.frame.text();
        if predicate(&text, &shot.frame) {
            return Ok(shot);
        }
        if Instant::now() >= deadline {
            bail!("{description}\n\nVisible screen:\n{text}");
        }
        thread::sleep(Duration::from_millis(10));
    }
}

fn wait_for_initial_screen(session: &mut Session) -> Result<Shot> {
    wait_for_screen(
        session,
        "TUI did not render its initial task view",
        STARTUP_TIMEOUT,
        |text, _| {
            text.contains("Tasks (/ - Search)")
                && text.contains("interactive#dev")
                && text.contains("finite#dev")
                && text.contains("INTERACTIVE_READY")
        },
    )
}

fn wait_for_transcript_idle(session: &mut Session) -> Result<Vec<u8>> {
    let deadline = Instant::now() + STARTUP_TIMEOUT;
    let mut transcript = session.logs(true)?;
    let mut quiet_since = Instant::now();
    loop {
        thread::sleep(Duration::from_millis(50));
        let next = session.logs(true)?;
        if next.len() != transcript.len() {
            transcript = next;
            quiet_since = Instant::now();
        } else if quiet_since.elapsed() >= Duration::from_millis(500) {
            return Ok(next);
        }
        if Instant::now() >= deadline {
            bail!("terminal output did not become idle");
        }
    }
}

fn add_overflowing_task_list(workspace: &Path) -> Result<()> {
    fs::write(
        workspace.join("turbo.json"),
        serde_json::to_vec(&json!({
            "$schema": "https://turborepo.dev/schema.json",
            "ui": "tui",
            "tasks": { "list": { "cache": false } }
        }))?,
    )?;

    for index in 0..20 {
        let name = format!("zz-task-{index:02}");
        let package_dir = workspace.join("packages").join(&name);
        fs::create_dir(&package_dir)?;
        fs::write(
            package_dir.join("package.json"),
            serde_json::to_vec(&json!({
                "name": name,
                "private": true,
                "scripts": { "list": "node -e \"setTimeout(() => {}, 30000)\"" }
            }))?,
        )?;
    }
    Ok(())
}

fn visible_sidebar_tasks(text: &str) -> Vec<&str> {
    text.lines()
        .filter_map(|line| line.split_once('│').map(|(sidebar, _)| sidebar))
        .flat_map(str::split_whitespace)
        .filter(|word| word.ends_with("#list"))
        .collect()
}

fn cells_for_text<'a>(frame: &'a Frame, text: &str) -> Result<Vec<&'a Cell>> {
    let characters = text.chars().collect::<Vec<_>>();
    for y in 0..frame.rows {
        for x in 0..=frame.cols.saturating_sub(characters.len() as u16) {
            let cells = characters
                .iter()
                .enumerate()
                .map(|(offset, character)| {
                    frame.cells.iter().find(|cell| {
                        cell.x == x + offset as u16
                            && cell.y == y
                            && cell.text.starts_with(*character)
                    })
                })
                .collect::<Option<Vec<_>>>();
            if let Some(cells) = cells {
                return Ok(cells);
            }
        }
    }
    bail!("visible frame did not contain {text:?}")
}

#[test]
fn renders_tasks_and_navigates_between_their_output() -> Result<()> {
    let mut context = launch_tui()?;
    wait_for_initial_screen(&mut context.session)?;

    context.session.send(b"\x1b[B")?;
    wait_for_screen(
        &mut context.session,
        "finite task output was not selected",
        STARTUP_TIMEOUT,
        |text, _| text.contains("FINITE_COMPLETE"),
    )?;

    context.session.send(b"\x1b[A")?;
    wait_for_screen(
        &mut context.session,
        "interactive task output was not reselected",
        STARTUP_TIMEOUT,
        |text, _| text.contains("INTERACTIVE_READY"),
    )?;
    Ok(())
}

#[test]
fn preserves_task_output_colors_end_to_end() -> Result<()> {
    let mut context = launch_tui()?;
    let shot = wait_for_screen(
        &mut context.session,
        "colored task output did not render",
        STARTUP_TIMEOUT,
        |text, _| text.contains("COLOR_OUTPUT"),
    )?;

    let cells = cells_for_text(&shot.frame, "COLOR_OUTPUT")?;
    assert!(cells.iter().all(|cell| {
        cell.foreground
            == Color {
                r: 12,
                g: 200,
                b: 123,
            }
    }));
    Ok(())
}

#[test]
fn drives_search_and_popup_overlays() -> Result<()> {
    let mut context = launch_tui()?;
    wait_for_initial_screen(&mut context.session)?;

    context.session.send(b"/finite")?;
    context.session.send(b"\r")?;
    wait_for_screen(
        &mut context.session,
        "search did not select the finite task",
        STARTUP_TIMEOUT,
        |text, _| text.contains("/ finite") && text.contains("FINITE_COMPLETE"),
    )?;

    context.session.send(b"\x1b")?;
    context.session.send(b"m")?;
    wait_for_screen(
        &mut context.session,
        "keybind popup did not open",
        STARTUP_TIMEOUT,
        |text, _| text.contains("Keybinds") && text.contains("Toggle log panel"),
    )?;

    context.session.send(b"\x1b")?;
    context.session.send(b"l")?;
    wait_for_screen(
        &mut context.session,
        "log panel did not open",
        STARTUP_TIMEOUT,
        |text, _| text.contains("l to close"),
    )?;
    Ok(())
}

#[test]
fn forwards_input_to_an_interactive_task_and_returns_to_navigation() -> Result<()> {
    let mut context = launch_tui()?;
    wait_for_initial_screen(&mut context.session)?;

    context.session.send(b"i")?;
    wait_for_screen(
        &mut context.session,
        "interactive mode did not start",
        STARTUP_TIMEOUT,
        |text, _| text.contains("Ctrl-z - Stop interacting"),
    )?;

    context.session.send(b"hello-from-test")?;
    context.session.send(b"\r")?;
    wait_for_screen(
        &mut context.session,
        "task did not receive interactive input",
        STARTUP_TIMEOUT,
        |text, _| text.contains("INTERACTIVE_ECHO:hello-from-test"),
    )?;

    context.session.send(b"\x1a")?;
    wait_for_screen(
        &mut context.session,
        "interactive mode did not return to navigation",
        STARTUP_TIMEOUT,
        |text, _| {
            text.contains("Tasks (/ - Search)") && !text.contains("Ctrl-z - Stop interacting")
        },
    )?;
    Ok(())
}

#[test]
fn streams_logs_and_restores_the_terminal_on_shutdown() -> Result<()> {
    let mut context = launch_tui()?;
    wait_for_initial_screen(&mut context.session)?;

    context.session.send(b"h")?;
    wait_for_screen(
        &mut context.session,
        "selected task logs did not start streaming",
        STARTUP_TIMEOUT,
        |text, _| {
            text.contains("Streaming logs for interactive#dev. Press h to return to the TUI.")
                && text.contains("INTERACTIVE_READY")
        },
    )?;

    context.session.send(b"h")?;
    wait_for_screen(
        &mut context.session,
        "TUI did not return after streaming",
        STARTUP_TIMEOUT,
        |text, _| {
            text.contains("INTERACTIVE_READY")
                && !text.contains("Streaming logs for interactive#dev")
        },
    )?;

    context.session.send(b"\x03")?;
    context
        .session
        .wait_for_exit(Duration::from_secs(10))?
        .context("turbo did not exit after Ctrl-C")?;
    let restored = capture(&mut context.session)?;
    assert!(!restored.frame.text().contains("Tasks (/ - Search)"));
    Ok(())
}

#[test]
#[ignore = "known bug: mouse-wheel input over the sidebar does not scroll the task list"]
fn scrolls_the_task_list_with_the_mouse_wheel_over_the_sidebar() -> Result<()> {
    let mut context = launch_tui_with(
        "list",
        &["--concurrency=10"],
        100,
        15,
        add_overflowing_task_list,
    )?;
    wait_for_screen(
        &mut context.session,
        "overflowing task list did not render",
        STARTUP_TIMEOUT,
        |text, _| text.contains("Tasks (/ - Search)") && visible_sidebar_tasks(text).len() >= 8,
    )?;
    wait_for_transcript_idle(&mut context.session)?;
    let before = capture(&mut context.session)?;
    let before_text = before.frame.text();
    let before_tasks = visible_sidebar_tasks(&before_text);

    context
        .session
        .send(b"\x1b[<65;5;5M\x1b[<65;5;5M\x1b[<65;5;5M\x1b[<65;5;5M\x1b[<65;5;5M")?;
    thread::sleep(Duration::from_millis(500));

    let after = capture(&mut context.session)?;
    assert_ne!(visible_sidebar_tasks(&after.frame.text()), before_tasks);
    Ok(())
}

#[test]
fn repaints_the_tui_immediately_after_a_terminal_resize() -> Result<()> {
    let mut context = launch_tui()?;
    wait_for_initial_screen(&mut context.session)?;
    let before = wait_for_transcript_idle(&mut context.session)?;

    for step in 0..=10 {
        context.session.resize(110 + step, 26 + step, 9, 18)?;
        thread::sleep(Duration::from_millis(10));
    }

    let deadline = Instant::now() + Duration::from_secs(2);
    loop {
        let after = context.session.logs(true)?;
        let screen = capture(&mut context.session)?;
        let repainted = after.get(before.len()..).is_some_and(|repaint| {
            repaint
                .windows(b"\x1b[2J".len())
                .any(|window| window == b"\x1b[2J")
        });
        if screen.frame.cols == 120 && screen.frame.rows == 36 && repainted {
            break;
        }
        if Instant::now() >= deadline {
            bail!("turbo did not clear and repaint the resized terminal");
        }
        thread::sleep(Duration::from_millis(10));
    }
    Ok(())
}
